### PR TITLE
Add suite to test

### PR DESCRIPTION
**Pull request by sergiors**

### CODE REVIEW

File: phpunit.xml.dist [hunk 1]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ikari7789) [inline phpunit.xml.dist:1]: Why leave this as ".dist?" Should this be the standard config for PHPUnit unit tests rather than an example?

**Comment by reisraff** [comment]
It should be added to .travis.yml also

**Comment by gabrielrcouto** [comment]
Waiting for some replies and adjustments to accept this merge. Good idea to use phpunit on this project, I hope we can implement tests on the core (after the refactor which I'm doing).

File: phpunit.xml.dist (hunk 1 above)
Comment (henriquemoody) [inline phpunit.xml.dist:1]: PHPUnit checks if _phpunit.xml.dist_ exists after checking if _phpunit.xml_ exists.

This allows you to have an standard XML (_phpunit.xml.dist_), but also that you can have a custom XML (_phpunit.xml_).

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,4 @@ material/
 vendor/
 roms/*rom
 cache.properties
+build/
```
Comment (henriquemoody) [inline .gitignore:4]: Please add `phpunit.xml` to the list.

File: phpunit.xml.dist [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="vendor/phpunit/phpunit/phpunit.xsd"
+         backupGlobals="false"
+         backupStaticAttributes="false"
+         colors="true"
```
Comment (henriquemoody) [inline phpunit.xml.dist]: You better remove that from the default phpunit.xml

File: phpunit.xml.dist [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        <whitelist>
+            <directory suffix=".php">src/</directory>
+            <exclude>
+                <directory suffix=".php">vendor/</directory>
+            </exclude>
+        </whitelist>
+    </filter>
+
+    <php>
+        <ini name="date.timezone" value="America/Sao_Paulo"/>
```
Comment (henriquemoody) [inline phpunit.xml.dist]: _UTC_ would be better, don't you think?

File: phpunit.xml.dist (hunk 3 above)
Comment (sergiors) [inline phpunit.xml.dist]: :+1: @henriquemoody has any other attrribute which I forgot to remove?

File: .gitignore (hunk 2 above)
Comment (sergiors) [inline .gitignore:4]: Done! Thanks!

File: phpunit.xml.dist (hunk 4 above)
Comment (sergiors) [inline phpunit.xml.dist]: Removed!

**Comment by sergiors** [comment]
@reisraff when ran phing:

```
[phingcall] /Users/sergio/PhpstormProjects/tests/php-terminal-gameboy-emulator/build.xml:21:51: Directory /Users/sergio/PhpstormProjects/tests/php-terminal-gameboy-emulator/${builddir}/src not found.
```

**Comment by reisraff** [comment]
@sergiors phing is not necessary to run these tests. Follow the last version of the build.xml before you change.

File: phpunit.xml.dist [hunk 5]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="vendor/phpunit/phpunit/phpunit.xsd"
+         backupGlobals="false"
```
Comment (henriquemoody) [inline phpunit.xml.dist]: @sergiors, IMO, only `bootstrap` and this one are enough.

I think you can let the rest as default with no problems but, differently from the comment about the `colors` attribute, this is just an opinion.

**Comment by reisraff** [comment]
@sergiors a example could be:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<project name="php-terminal-gameboy-emulator" default="build">
    <property name="vendordir" value="${basedir}/bin" />
    <target name="check" depends="lint,phpcs,phpunit"/>

    <target name="lint" description="Perform syntax check of sourcecode files">
        <apply executable="php" failonerror="true">
            <arg value="-l" />

            <fileset dir="${basedir}/src">
                <include name="**/*.php" />
                <modified />
            </fileset>
        </apply>
    </target>

    <target name="phpcs" description="Find coding standard violations using PHP_CodeSniffer and print human readable output. Intended for usage on the command line before committing.">
        <exec executable="${vendordir}/phpcs" failonerror="true">
            <arg value="--extensions=php" />
            <arg value="--standard=${basedir}/phpcs.xml" />
            <arg path="${basedir}/src" />
        </exec>
    </target>

    <target name="phpunit" description="Run unit tests with PHPUnit">
        <exec executable="${vendordir}/phpunit" failonerror="true">
            <arg value="-c" />
            <arg path="${basedir}" />
        </exec>
    </target>
</project>
```

**Comment by sergiors** [comment]
I think it would be better Phing.
Once you control with just PHP.
and could add `phploc/phploc, phpmd/phpmd, sebastian/phpcpd, squizlabs/php_codesniffer, pdepend/pdepend`...

Ant depends of me install on my OS

and build.xml like a this:

```
<?xml version="1.0" encoding="UTF-8"?>
<project name="php-terminal-gameboy-emulator" default="build" basedir=".">
    <property name="project.bindir" value="${project.basedir}/vendor/bin" />
    <property name="project.builddir" value="${project.basedir}/build" />

    <target name="build">
        <phingcall target="lint" />
        <phingcall target="tests" />
        <phingcall target="style" />
    </target>

    <target name="style">
        <exec executable="${project.bindir}/phpcs" passthru="true" checkreturn="true" >
            <arg value="--standard=PSR2" />
            <arg value="${project.basedir}/src" />
            <arg value="${project.basedir}/tests" />
        </exec>
    </target>

    <target name="tests">
        <exec command="${project.bindir}/phpunit --testsuite unit" passthru="true" checkreturn="true" />
    </target>

    <target name="lint">
        <phplint haltonfailure="true">
            <fileset dir="${project.basedir}/src">
                <include name="**/*.php" />
            </fileset>
            <fileset dir="${project.basedir}/tests">
                <include name="**/*.php" />
            </fileset>
        </phplint>
    </target>

    <target name="qa" depends="prepare-ci">
        <phingcall target="tests-ci" />
        <phingcall target="loc-ci" />
        <phingcall target="dependencies-ci" />
        <phingcall target="phpmd-ci" />
        <phingcall target="style-ci" />
        <phingcall target="duplications-ci" />
    </target>

    <target name="prepare-ci">
        <delete dir="${project.builddir}/logs" />
        <delete dir="${project.builddir}/coverage" />
        <delete dir="${project.builddir}/images" />

        <mkdir dir="${project.builddir}/logs" mode="0777" />
        <mkdir dir="${project.builddir}/coverage" mode="0777" />
        <mkdir dir="${project.builddir}/images" mode="0777" />
    </target>

    <target name="tests-ci">
        <exec executable="${project.bindir}/phpunit" passthru="true">
            <arg value="--coverage-html" />
            <arg value="${project.builddir}/coverage" />
            <arg value="--coverage-clover" />
            <arg value="${project.builddir}/logs/clover.xml" />
            <arg value="--log-junit" />
            <arg value="${project.builddir}/logs/junit.xml" />
        </exec>
    </target>

    <target name="loc-ci">
        <exec executable="${project.bindir}/phploc" passthru="true">
            <arg value="--count-tests" />
            <arg value="--log-csv" />
            <arg value="${project.builddir}/logs/phploc.csv" />
            <arg value="--log-xml" />
            <arg value="${project.builddir}/logs/phploc.xml" />
            <arg value="${project.basedir}/src" />
            <arg value="${project.basedir}/tests" />
        </exec>
    </target>

    <target name="dependencies-ci">
        <exec executable="${project.bindir}/pdepend" passthru="true">
            <arg value="--jdepend-xml=${project.builddir}/logs/jdepend.xml" />
            <arg value="--jdepend-chart=${project.builddir}/images/dependencies.svg" />
            <arg value="--overview-pyramid=${project.builddir}/images/overview-pyramid.svg" />
            <arg path="${project.basedir}/src" />
        </exec>
    </target>

    <target name="phpmd-ci">
        <exec executable="${project.bindir}/phpmd" passthru="true">
            <arg path="${project.basedir}/src" />
            <arg value="xml" />
            <arg value="cleancode,codesize,controversial,design,naming,unusedcode" />
            <arg value="--reportfile" />
            <arg value="${project.builddir}/logs/pmd.xml" />
        </exec>
    </target>

    <target name="style-ci">
        <exec executable="${project.bindir}/phpcs" passthru="true">
            <arg value="--report=checkstyle" />
            <arg value="--report-file=${project.builddir}/logs/checkstyle.xml" />
            <arg value="--standard=PSR2" />
            <arg value="${project.basedir}/src" />
            <arg value="${project.basedir}/tests" />
        </exec>
    </target>

    <target name="duplications-ci">
        <exec executable="${project.bindir}/phpcpd" passthru="true">
            <arg value="--log-pmd" />
            <arg value="${project.builddir}/logs/pmd-cpd.xml" />
            <arg path="${project.basedir}/src" />
        </exec>
    </target>
</project>
```

**Comment by henriquemoody** [comment]
I think this is really a matter of taste... 

I'm in favor of use Composer to run these scripts, since the use of composer is _kinda of_ mandatory to fetch the dependencies, then you don't have to install Phing, Ant, Make, etc...

---

_Edit_: However, this is something to discuss in another issue/pull request. For this one I think you should stick with Ant.

**Comment by reisraff** [comment]
@henriquemoody I agree with you about to test with composer, but the code being that way is just modify. So @sergiors an example to test everything:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<project name="php-terminal-gameboy-emulator" default="build">
    <property name="vendordir" value="${basedir}/bin" />
    <target name="build" depends="prepare,lint,phploc,pdepend,phpmd-ci,phpcs-ci,phpcpd,phpunit-ci"/>
    <target name="build-parallel" depends="prepare,lint,tools-parallel,phpunit-ci"/>
    <target name="check" depends="lint,phpcs,phpunit"/>

    <target name="tools-parallel" description="Run tools in parallel">
        <parallel threadCount="2">
            <sequential>
                <antcall target="pdepend"/>
                <antcall target="phpmd-ci"/>
            </sequential>
            <antcall target="phpcpd"/>
            <antcall target="phpcs-ci"/>
            <antcall target="phploc"/>
        </parallel>
    </target>

    <target name="clean" description="Cleanup build artifacts">
        <delete dir="${basedir}/build/coverage"/>
        <delete dir="${basedir}/build/logs"/>
        <delete dir="${basedir}/build/pdepend"/>
    </target>

    <target name="prepare" depends="clean" description="Prepare for build">
        <mkdir dir="${basedir}/build/coverage"/>
        <mkdir dir="${basedir}/build/logs"/>
        <mkdir dir="${basedir}/build/pdepend"/>
        <exec executable="composer" failonerror="true">
            <arg value="install" />
        </exec>
    </target>

    <target name="lint" description="Perform syntax check of sourcecode files">
        <apply executable="php" failonerror="true">
            <arg value="-l" />

            <fileset dir="${basedir}/src">
                <include name="**/*.php" />
                <modified />
            </fileset>
        </apply>
    </target>

    <target name="phploc" description="Measure project size using PHPLOC">
        <exec executable="${vendordir}/phploc">
            <arg value="--count-tests" />
            <arg value="--log-csv" />
            <arg value="${basedir}/build/logs/phploc.csv" />
            <arg path="${basedir}/src" />
        </exec>
    </target>

    <target name="pdepend" description="Calculate software metrics using PHP_Depend">
        <exec executable="${vendordir}/pdepend">
            <arg value="--jdepend-xml=${basedir}/build/logs/jdepend.xml" />
            <arg value="--jdepend-chart=${basedir}/build/pdepend/dependencies.svg" />
            <arg value="--overview-pyramid=${basedir}/build/pdepend/overview-pyramid.svg" />
            <arg path="${basedir}/src" />
        </exec>
    </target>

    <target name="phpmd" description="Perform project mess detection using PHPMD and print human readable output. Intended for usage on the command line before committing.">
        <exec executable="${vendordir}/phpmd">
            <arg path="${basedir}/src" />
            <arg value="text" />
            <arg value="${basedir}/app/phpmd.xml" />
        </exec>
    </target>

    <target name="phpmd-ci" description="Perform project mess detection using PHPMD creating a log file for the continuous integration server">
        <exec executable="${vendordir}/phpmd">
            <arg path="${basedir}/src" />
            <arg value="xml" />
            <arg value="${basedir}/app/phpmd.xml" />
            <arg value="--reportfile" />
            <arg value="${basedir}/build/logs/pmd.xml" />
        </exec>
    </target>

    <target name="phpcs" description="Find coding standard violations using PHP_CodeSniffer and print human readable output. Intended for usage on the command line before committing.">
        <exec executable="${vendordir}/phpcs">
            <arg value="--extensions=php" />
            <arg value="--standard=${basedir}/app/phpcs.xml" />
            <arg path="${basedir}/src" />
        </exec>
    </target>

    <target name="phpcs-ci" description="Find coding standard violations using PHP_CodeSniffer creating a log file for the continuous integration server">
        <exec executable="${vendordir}/phpcs">
            <arg value="--report=checkstyle" />
            <arg value="--extensions=php" />
            <arg value="--report-file=${basedir}/build/logs/checkstyle.xml" />
            <arg value="--standard=${basedir}/app/phpcs.xml" />
            <arg path="${basedir}/src" />
        </exec>
    </target>

    <target name="phpcpd" description="Find duplicate code using PHPCPD">
        <exec executable="${vendordir}/phpcpd">
            <arg value="--log-pmd" />
            <arg value="${basedir}/build/logs/pmd-cpd.xml" />
            <arg path="${basedir}/src" />
        </exec>
    </target>

    <target name="phpunit" description="Run unit tests with PHPUnit">
        <exec executable="${vendordir}/phpunit" failonerror="true">
            <arg value="-c" />
            <arg path="${basedir}/app" />
        </exec>
    </target>

    <target name="phpunit-ci" description="Run unit tests with PHPUnit">
        <exec executable="${vendordir}/phpunit" failonerror="true">
            <arg value="-c" />
            <arg path="${basedir}/app/phpunit.xml.ci" />
        </exec>
    </target>
</project>
```

and change the .travis.yml to execute `ant build`

**Comment by sergiors** [comment]
I see no advantage in maintaining an external tool when we have one.

**Comment by henriquemoody** [comment]
> I see no advantage in maintaining an external tool when we have one.

I'm with you on that. On my projects I don't using any tool, I just add the targets to the continuous integration service (not Composer, not Phing, not Ant, not Make, not anything). But as I said, that's a matter of taste.

**Comment by reisraff** [comment]
See #34 PR

**Comment by henriquemoody** [comment]
As I see this pull request is good to merge and must be merged before #34.

**Comment by henriquemoody** [comment]
@sergiors, can you squash these commits at once?

**Comment by sergiors** [comment]
@henriquemoody done.

**Comment by henriquemoody** [comment]
@sergiors, you may want to remove 28ff3c2 since it's already in master branch.

It seems master suffered from a forced push, would be nice if you rebase with the current master and do a forced push on you branch.

**Comment by sergiors** [comment]
@henriquemoody I don't to do this right 😂
I never really used rebase

File: phpunit.xml.dist (hunk 1 above)
Comment (ikari7789) [inline phpunit.xml.dist:1]: I don't agree with that at all. This project isn't some base project for people to base other projects off of like a framework. It's a single, definitive project that should have its own standards and not allow user's choice as the default. If someone wants to use their own settings they can modify the phpunit.xml in their own local checkout.

Comment (henriquemoody) [inline phpunit.xml.dist:1]: >  If someone wants to use their own settings they can modify the phpunit.xml in their own local checkout.

That's the whole point of having phpunit.xml.dist and having phpunit.xml on the .gitignore list.

Can you elaborate more?

Comment (ikari7789) [inline phpunit.xml.dist:1]: Certainly. If this were a framework or some other project meant for others to use as a base to start their own projects, then I can understand leaving phpunit.xml.dist, but this is not such a project. The .dist version is meant to be used as a template for you to quickly make your own setting file, but not necessarily use it as a default. It's misleading to a developer in my opinion. To me it says "you can run tests if you want, but we don't care enough to define our own test setup, so here's a file to get you started."

Comment (gabrielrcouto) [inline phpunit.xml.dist:1]: Well, maybe the .dist is a way to test some local configs whitout the need of discarding the config file on every commit. This is the point @henriquemoody? Why we will use a different config than the original?

Comment (gabrielrcouto) [inline phpunit.xml.dist:1]: Wow, comments at same time! Nice point @ikari7789, I agree.

**Comment by gabrielrcouto** [comment]
About the ant Versus phing: This project is about what PHP can do, will be nice to see Phing here than ant 😉 @sergiors, can you replicate the ant tasks on phing, including phpcs for travis? I already talked with @reisraff about this change.

File: phpunit.xml.dist (hunk 1 above)
Comment (henriquemoody) [inline phpunit.xml.dist:1]: The idea of _phpunit.xml.dist_ is to define the standard but allow people to customize. It's simple, some developers like `colors="true"`, `verbose="true"`, `stopOnFailure="true"`, etc, and some don't. With this _schema_ they can customize their own way to run PHPUnit.

Who uses _phpunit.xml.dist_?
- [Symfony](https://github.com/symfony/symfony/blob/master/phpunit.xml.dist);
- [Doctrine2](https://github.com/doctrine/doctrine2/blob/master/phpunit.xml.dist);
- [Monolog](https://github.com/Seldaek/monolog/blob/master/phpunit.xml.dist);
- [PHPunit](https://github.com/sebastianbergmann/phpunit/blob/master/phpunit.xml.dist) itself.

I don't think I can list all but just trying would be tiring. The projects I listed are not _base_ for anything, they just want that the developers who want to contribute to be comfortable about how they want to run the test suite.

Comment (gabrielrcouto) [inline phpunit.xml.dist:1]: Good point @henriquemoody, let's go on that way (with .dist) 😉

**Comment by ikari7789** [comment]
Not even really sure this needs neither tool. Just use the Travis CI config or add a test script to Composer.

**Comment by gabrielrcouto** [comment]
Hmm, the test script on composer is not a bad idea, I liked! More people will know how to run composer test than phing (I guess). What you thing guys?

**Comment by henriquemoody** [comment]
:+1: for Composer scripts

**Comment by reisraff** [comment]
:+1: for composer scripts

**Comment by sergiors** [comment]
Seriously I'll have to do all the work again?

**Pull request closed without merging**